jax/interpreters/batching.py

```diff
@@ -129,14 +132,29 @@ def process_primitive(self, primitive, tracers, params):
     vals_in, dims_in = unzip2((t.val, t.batch_dim) for t in tracers)
     if all(bdim is not_mapped for bdim in dims_in):
       return primitive.bind(*vals_in, **params)
+    elif config.omnistaging_enabled and primitive in collective_rules:
+      axes_names = params['axis_name']
+      if not isinstance(axes_names, (tuple, list)):
+        axes_names = (axes_names,)
+      for i, axis_name in enumerate(axes_names):
+        frame = core.axis_frame(axis_name)
+        if frame.tag is self.master:
```

Comment:
I know you explained this to me before, but I forget: is the tag just to handle possible shadowing? If so, after this came up in a convo with James and Roy last week, I've started to think we should make shadowing illegal, which might also help us slightly simplify this logic.

For now, just flagging this for more discussion!

Reply:
Yes, the tag is to prevent shadowing, and I think that based on our conversation we shouldn't be that eager to disallow it, right?

Reply:
Yeah, I agree.